flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java

```diff
@@ -99,44 +101,56 @@ public BlobCacheSizeTracker(long sizeLimit) {
     }
 
     /** Register the target file to the tracker. */
-    public void track(@Nullable JobID jobId, BlobKey blobKey, long size) {
+    public void track(JobID jobId, BlobKey blobKey, long size) {
+        checkNotNull(jobId);
         checkNotNull(blobKey);
         checkArgument(size >= 0);
 
         synchronized (lock) {
-            caches.put(Tuple2.of(jobId, blobKey), size);
-            if (jobId != null) {
+            if (caches.putIfAbsent(Tuple2.of(jobId, blobKey), size) == null) {
                 blobKeyByJob.computeIfAbsent(jobId, ignore -> new HashSet<>()).add(blobKey);
-            }
-            total += size;
-            if (total > sizeLimit) {
+
+                total += size;
+                if (total > sizeLimit) {
+                    LOG.warn(
+                            "The overall size of ShuffleDescriptors in PermanentBlobCache exceeds "
+                                    + "the limit. Limit = [{}], Current: [{}], "
+                                    + "The size of next ShuffleDescriptors: [{}].",
+                            sizeLimit,
+                            total,
+                            size);
+                }
+            } else {
                 LOG.warn(
```

Comment:
Is it possible to happen frequently. If yes, I prefer to make it a debug log.

Reply:
I think this warning shouldn't happen frequently. In `BlobUtil.moveTempFileToStore`, if the blob key duplicates, a warning will be outputted, too. Furthermore, the previous warning ("The overall size of ShuffleDescriptors in PermanentBlobCache exceeds ...") should be warning, too. The overall size exceeds if and only if the size of blob exceeds 100MiB. In this scenario the user should be warned about it.